### DISCUSSION TITLE
RE: debugging tree-edit

### DISCUSSION THREAD

**Discussion by ethan-leba** [ethan-leba/tree-edit#48] [General]
Wow exwm crashed while writing this comment and trying to reproduce this, but I have to write it again basically what happens is I start with code like that:

```elisp
def test:
    if TRUE:
        [X]print ("foo")
    print ("bar")
```
with the Cursor at [X] using evil-tree-edit-slurp I get that error:
`tree-edit-slurp: Wrong type argument: user-ptrp, nil`

Also I get that:
`user-error: Tree-edit does not support exwm-mode!`

If I deactivate xah-fly-keys I can theoreticillay use ">" I think I have to press it 2 times with maybe ESC before then instead of slurping the second print statement into the if block, it moves the 1. print statement a tab / 4 spaces to the right.

Why does tree-edit not support exwm-mode? What has exwm to do with formating of text buffers?

_Originally posted by @spiderbit in https://github.com/ethan-leba/tree-edit/issues/40#issuecomment-997274607_

**Comment by ethan-leba** [comment]
> Wow exwm crashed while writing this comment and trying to reproduce this, but I have to write it again basically what happens is I start with code like that:
> 
> ```emacs-lisp
> def test:
>     if TRUE:
>         [X]print ("foo")
>     print ("bar")
> ```
> 
> with the Cursor at [X] using evil-tree-edit-slurp I get that error: `tree-edit-slurp: Wrong type argument: user-ptrp, nil`

Are you in tree state when calling this command (`M-x evil-tree-state`)? There needs to be a current node selected for the command to execute. The current node is indicated by a highlighted region as seen in the GIF.

> Also I get that: `user-error: Tree-edit does not support exwm-mode!`

Hmm, is `exwm-mode` a major mode? `evil-tree-edit` will try to load a config file based on the major mode, and it's not finding it there.

> If I deactivate xah-fly-keys I can theoreticillay use ">" I think I have to press it 2 times with maybe ESC before then instead of slurping the second print statement into the if block, it moves the 1. print statement a tab / 4 spaces to the right.

This is likely evil's normal state binding (`evil-shift-right`).

For integrating into your own modes you may be interested in this [example](https://github.com/ethan-leba/tree-edit/blob/main/doc/using-tree-edit.org#emacs-style) 

These commands could definitely have better error messages to indicate what's going on here, thanks for bringing this to my attention!

**Comment by spiderbit** [comment]
Ok I can use evil-tree-state with Q and then > and then it says: nothing to slurp. the "if" block is marked, but it does not slurp the "bar" line.

**Reply by ethan-leba** [reply]
It's a bit unintuitive, but `tree-edit` expects you to have the block selected, since that's you're slurping into:

```
if foo:
    [bar]
baz

to

if foo:
    [bar
    baz]
```
